### ISSUE TITLE
Tutorial on Query Loop block variations

### ISSUE THREAD

**Issue by justintadlock** [WordPress/developer-blog-content#39]
### Discussed in https://github.com/WordPress/developer-blog-content/discussions/34

<sup>Originally posted by **justintadlock** December  9, 2022</sup>
I'd like to put together a foundational tutorial that covers building custom Query Loop variations for plugin authors.  At its most basic, this can be something as simple as using the built-in allowed controls and inner blocks for the template.  

However, it's probably worth exploring some basic post metadata inclusion with a custom settings control for filtering the query.  For example, if showing a "book reviews" variation, the tutorial could also teach how to filter those posts by a "star rating" custom field.

I have most of this mapped out in my mind at this point, but any feedback on what anyone would like to see the tutorial cover is welcome.

**Comment by justintadlock** [comment]
To get this tutorial ready, I first needed to create an example plugin that would be easy enough to walk developers through without getting into everything that could possibly go along with this topic.

Plugin link: https://github.com/justintadlock/book-review-variations

What I settled on was showing how a developer could build a custom "Book Reviews" variation (potentially for a client) using the standard WP "post" post type (hat tip to @bph for some earlier discussion around this).  Since I'm using posts, there's no need to worry about setup around a CPT.

The goal of the post will be to walk through setting up a simple Query Loop block variation in the first step.  The second step of the tutorial will be integrating the query with a custom field (star rating) and block control to go with it.

Here's a screenshot of it in action:

![book-reviews-001](https://user-images.githubusercontent.com/1816309/207160749-36705614-7485-4e3e-bd1b-41a1ad99e197.jpeg)

I expect to have a draft of this ready in the next day.

**Comment by justintadlock** [comment]
So, I think I bit off more than I could chew for a single day.  I'll definitely have a draft ready for review tomorrow.

Just uploading a backup copy of my unfinished draft for now: [Query Loop variations early draft](https://github.com/WordPress/developer-blog-content/files/10222647/index.md)

**Comment by justintadlock** [comment]
The first draft is now [available on Google Docs](https://docs.google.com/document/d/1icrO-KQDMMg5awA-M8Di9AhXxg_2VsaxSkRH7CAha1E/edit?usp=sharing).  It is not quite ready for a first "official" review, but it is open to feedback at this point for anyone who wants to chime in.

I plan to go through a copyediting phase tomorrow to polish this draft and prep screenshots.  Then, it will be ready to go through the review process.

I'm also a bit concerned about length.  Currently, the post is over 2,200 words (counting code examples).  There's probably a section I could cut on setup, but if we feel like this is too long, I wrote the post in a way that it could also be broken up into a two-part tutorial.  There are two distinct major sections (simple variation and advanced variation additions).

**Comment by bph** [comment]
Great work! I built the query block variation following your article. It all works. 
Didn't work at first. I had to debug it. Found a typo in **my** code 🤦‍♀️ 
Left a couple of comments, too.
I am looking forward to read the final version :-)

**Comment by justintadlock** [comment]
Thanks for going through this @bph, especially for running through the code itself.  That's often the most tedious part of the process, so it's a big help knowing that you got it working.

**Comment by mburridge** [comment]
Great article. Can confirm that all the code works. I also encountered a couple of hiccups along the way but managed to resolve them. I've added a few comments to the doc.

**Comment by justintadlock** [comment]
@mburridge Thanks for the great feedback.  That was super helpful in wrangling this thing.

On that note, are we able to upload a WordPress export XML file for readers to download and import into their sites?  @mburridge had the idea of making the content available to import to cut back on some of the setup.

---

The post is now on the Developer Blog and ready to go through the review process.  

- [Preview link](https://developer.wordpress.org/news/?p=530&preview=true).
- [Public preview link](https://developer.wordpress.org/news/?p=530&preview=1&_ppp=9f90926f11) for those without admin access.

It's a bit hefty, so multiple sets of eyes would be appreciated.

**Comment by juanmaguitar** [comment]
@justintadlock I'm missing a link to a repo with the code shown in the article
https://github.com/justintadlock/book-review-variations

I would also prepare the repo so it can be quickly seen in action by adding `wp-env` to the dependencies and some scripts to launch a Local WordPress w/ the plugin already installed

Some references for this:
- https://github.com/wordpress-juanmaguitar/demo-plugin-wp-data
- https://github.com/wordpress-juanmaguitar/demo-block-wp-env

**Comment by mburridge** [comment]
> are we able to upload a WordPress export XML file for readers to download and import into their sites

> I'm missing a link to a repo with the code shown in the article

The repo could contain the .xml file and image files as well.

**Comment by justintadlock** [comment]
@juanmaguitar The repo is not a part of the tutorial.  I was using it to test out the code/ideas for what would eventually make it into the tutorial.

It's certainly possible to link to the repo, and I'm OK with that.  But, we really need an official home to house such code rather than personal GitHub accounts.  Currently, there is no decision on that, and it has not been standard to link to GitHub code examples on the Dev Blog so far.

**Comment by abhansnuk** [comment]
Hello, really thorough tutorial and it will be very useful. Thanks for tackling it @justintadlock .

Can we break it up into two as the discussion above, or add links to help readers return to the table of contents?

I have had non-English speakers query meaning of 'ton' before, so perhaps use an alternative phrase in the intro.
Suggest adding 'release' after '6.1' in the intro. I know for most it will be understood but may be useful for those who don't know the numbering and for search.

There seems to be double spaces between sentences. From memory, I think we are using one space.

I really like this phrase "The Query Loop block is a workhorse in building websites out of blocks." Great way to summarize it and would work well for social media too. Par 4 really gives extenders a reason to use it too. Wonder if we can say something about skipping block development in the intro too.

Third par: "were often building entirely custom blocks on their own." Suggest 'entire'.

Fifth par: "A small sampling of the ways this can be used include:" Suggest "A few examples of the way this can be used include" to make it easier for those translating.
The bullet points - may need to change to the style we are using. Will need to look up what we landed on for this.

Sixth par: "For this walk-through, you will learn how to tackle the last item in that list." I know it is above, but for some readers and if reading on a small screen or through a screen reader, it may be easier to say what the item is.

Seventh par: "To keep this tutorial simple, I didn’t want developers to have a lot of legwork to do before stepping into it. However, the basic methods that you will learn can be applied to more complex projects." Suggest "The basic methods in this simple tutorial can be applied to more complex projects." Unless you were intending to add something after the first sentence.

Requirements: - don't think the colon is needed as it is a heading. Can it be added to the heading hierarchy if it is not already please.

"For further information on setting these up, ..." can we change the 'see' to something like 'go to' or 'visit' as some people will be listening to this post,

Table of Contents heading - is it possible to add the word 'Tutorial' to this heading so readers know this is the start of the tutorial itself. Will add comments on the tutorial below to avoid this message becoming too long.

**Comment by abhansnuk** [comment]
Suggestions on the Tutorial section @justintadlock .

**Building a simple Query Loop variation**
- "Now, it is time to register the variation. First, add a few basic properties, such as the name, title, and more:" can we add something to indicate as in the code block below. 

- "For this tutorial, the variation displays the latest six posts within the “Book Reviews” category. It also has a wide layout in a three-column grid. Feel free to tinker with the various options." - "Tinker" has been a word that I have heard from others does not translate easily.

- "For a full list of controls and their definitions, see [allowed controls section](https://developer.wordpress.org/block-editor/how-to-guides/block-tutorial/extending-the-query-loop-block/#disabling-irrelevant-or-unsupported-query-controls) in the developer docs." and "For a full overview of the available options, see following resources:"
Suggest we try and avoid lots of references to 'see' where they can be easily changed to another word, as some readers will be listening to this post.

For the second sentence (listed above), can we say these are in the Block Editor Handbook to promote that one exists for those who have not come across it before? It also tells people where they are going to for these links.

"You could stop at this point if your variation doesn’t require any customizations to the queried posts beyond what the core Query Loop block handles out of the box." Suggest adding 'of the tutorial' after 'at this point'.

**PHP section**
"If it sounds a little convoluted, it is. Ideally, there will be a slightly less complex method for doing this in the future." Is it worth checking with @annezazu or someone else in FSE if there are plans on this or feedback they are seeking in this area?

**Comment by abhansnuk** [comment]
In the intro, shall we link to the Query Loop block documentation / explanation in core?

For tags, how about adding:
'Query Loop', 'Custom Post Type'

@bph has had an answer on code blocks titles from the accessibility team.

Any there any links to Learn WordPress that would be useful, or suggestions for Learn WordPress materials that could link to this tutorial once it is live?

Before publishing, contributors/ those who have reviewed, plus an excerpt will need to be added to the post.

Great tutorial Justin.

**Comment by bph** [comment]
I posted the discussion around the code block on my personal blog. And if you are interested go there. [Recommendation for  content creators](https://icodeforapurpose.com/accessibility-aspects-on-using-code-block/#3-recommendation-for-writers-on-the-dev-blog)

the TL:DR is - For Content creators: 
>In general code block titles are not required.
>They are only needed when there are multiple code examples within the same section, (under one heading)  of a post

**Comment by justintadlock** [comment]
Thanks @abhansnuk for the feedback.  That was super helpful.  I've incorporated most of the changes and will do a final round tomorrow.

I think by the end of this one I'll be able to remember to say "visit X" instead of "see X". 😀 

> Can we break it up into two as the discussion above, or add links to help readers return to the table of contents?

Do we have a standard way to return to the ToC?  It's definitely something to consider. I just hit the "back" button in my browser. :) 

> Requirements: - don't think the colon is needed as it is a heading. Can it be added to the heading hierarchy if it is not already please.

Removed the colon.  I also made it a proper heading and adjusted its style.

@bph and I have talked about setting up a custom block style/pattern for callouts/notes/warnings/etc.  That's what we really need for sections like this to keep them consistent across the site.

> For tags, how about adding:
> 'Query Loop', 'Custom Post Type'

I'm OK with that.  

Do we have any guidelines on how we are tagging stuff?  For example, I've worked on projects with thousands of tags, each with only a single post, so the tags were essentially useless for finding related content.  It was a bit of a management nightmare down the road.  I worry that we might end up in a similar situation if we get too specific with tags early on.  This is not necessarily something we need to answer right away but maybe something to keep in mind as we go forward.
